IdentityCore/src/throttling/thumbprint_calculation/MSIDSilentRequestThumbprintCalculator.m

```diff
@@ -0,0 +1,156 @@
+
+@end
+
+//Exclude List:
+//1) Client ID - same across all requests
+//2) Grant type - fixed as @"refresh_token"
+@implementation MSIDSilentRequestThumbprintCalculator
+
+- (instancetype)initWithParamaters:(NSDictionary *)parameters
+                       endpointUrl:(NSString *)endpointUrl
+                             realm:(NSString *)realm
+                       environment:(NSString *)environment
+                     homeAccountId:(NSString *)homeAccountId
+{
+    self = [super init];
+    if (self)
+    {
+        _requestParameters = [parameters mutableCopy];
+        _requestParameters[@"endpointUrl"] = endpointUrl;
+        _requestParameters[@"realm"] = realm;
+        _requestParameters[@"environment"] = environment;
+        _requestParameters[@"homeAccountId"] = homeAccountId;
+        _strictThumbprintIncludeSet = [NSSet setWithArray:@[@"realm",@"environment",@"homeAccountId",MSID_OAUTH2_SCOPE]];
+        _fullThumbprintExcludeSet = [NSSet setWithArray:@[MSID_OAUTH2_CLIENT_ID,
+                                                          MSID_OAUTH2_GRANT_TYPE]];
+    }
+    return self;
+}
+
+- (NSString *)getFullRequestThumbprint
+{
+    NSString *fullRequestThumbprintKey = [self getRequestThumbprintImpl:self.fullThumbprintExcludeSet
+                                                        comparePolarity:NO];
+    if (!fullRequestThumbprintKey)
+    {
+        //Log Error
+        return nil;
+    }
+    return fullRequestThumbprintKey;
+}
+
+- (NSString *)getStrictRequestThumbprint
+{
+    NSString *strictRequestThumbprintKey = [self getRequestThumbprintImpl:self.strictThumbprintIncludeSet
+                                                          comparePolarity:YES];
+    if (!strictRequestThumbprintKey)
+    {
+        //Log Error
+        return nil;
+    }
+    return strictRequestThumbprintKey;
+}
+
+- (NSString *)getRequestThumbprintImpl:(NSSet *)filteringSet
+                       comparePolarity:(BOOL)comparePolarity
+{
+    NSArray *sortedThumbprintRequestList = [self sortRequestParametersUsingFilteredSet:filteringSet
+                                                                       comparePolarity:comparePolarity];
+    if (sortedThumbprintRequestList)
+    {
+        NSUInteger thumbprintKey = [self hash:sortedThumbprintRequestList];
+        if (thumbprintKey == 0)
+        {
+            return nil;
+        }
+        
+        else
+        {
+            return [NSString stringWithFormat:@"%lu", thumbprintKey];
+        }
+    }
+    return nil;
+}
+
+- (NSArray *)sortRequestParametersUsingFilteredSet:(NSSet *)filteringSet
+                                   comparePolarity:(BOOL)comparePolarity
+{
+    NSMutableArray *arrayList = [NSMutableArray new];
+    [self.requestParameters enumerateKeysAndObjectsUsingBlock:^(id  _Nonnull key, id  _Nonnull obj, __unused BOOL * _Nonnull stop) {
```

Comment:
This code would need to be repeated in each calculator. I'd recommend creating a generic calculator that takes a dict or a list of parameters, sorts, and returns hash. Request class itself could return list of parameters that need to be accounted for in the thumbprint calculation, or the thumbprint itself.

Reply:
It's not repeated in each calculator. SSOSilentRequestThumbprintCalculator is a child of SilentRequestThumbprintCalculator, and it uses the API in the parent class without duplicating the code. This sort API is private, but available to the child SSO calculator class through categorization.